Dependency graph prerequisites

bjoeris opened the pull request:
These are a number of bug fixes and changes which are necessary for, but independent of the dependency graph work.

pmuetschard commented:
Please drop `Bazel: Disable binary stripping when building under "-c dbg"` This is already done at head anyways.

bjoeris commented:
@pmuetschard Done

AWoloszyn commented on gapis/api/vulkan/api/queue.api:
```diff
@@ -164,6 +171,10 @@ cmd VkResult vkQueueSubmit(
   leaveSubcontext()
   execPendingCommands(queue)
   fence
+  if (fence != 0) {
```
I am quite surprised this doesn't collide with the fence keyword above.

bjoeris replied:
Added some comments for the the next time someone has to read this confusing overloading of 'fence'.

bjoeris commented on gapis/api/vulkan/api/queue.api:
```diff
@@ -173,6 +184,7 @@ cmd VkResult vkQueueSubmit(
 cmd VkResult vkQueueWaitIdle(
     VkQueue queue) {
   if !(queue in Queues) { vkErrorInvalidQueue(queue) }
+  Queues[queue] = Queues[queue]
```
This is to make vkQueueWaitIdle read and write the relevant queue. Without this, vkQueueWaitIdle doesn't touch the state, will have no dependencies, and will be dropped by the new DCE.

Qining replied:
Nice, nice if we have a comment here.

bjoeris replied:
Upon further reflection, this is not always adequate to find dependencies on vkQueueWaitIdle. Reverting for now, will address in the dependency graph CL.

Qining commented on gapis/api/vulkan/api/renderpass_framebuffer.api:
```diff
@@ -344,3 +344,22 @@ cmd void vkCmdEndRenderPass(
+    switch desc.storeOp {
+      case VK_ATTACHMENT_STORE_OP_STORE: {
+        if desc.initialLayout != desc.finalLayout {
+          transitionImageLayout(attachment.Image, attachment.SubresourceRange, VK_IMAGE_LAYOUT_UNDEFINED, desc.finalLayout)
```
So if the `storeOp` is not `VK_ATTACHMENT_STORE_OP_STORE`, can we skip the layout transition?

AWoloszyn replied:
We should do the layout transition regardless, but only write the image subresource if we are storing the image.

bjoeris replied:
storeImageAttachment now transitions the image layout regardless of the storeOp.

ben-clayton commented on gapis/api/vulkan/errors.api:
```diff
@@ -43,6 +43,7 @@ extern void vkErrNullPointer(string pointerType)
 extern void vkErrUnrecognizedExtension(string name)
 extern void vkErrExpectNVDedicatedlyAllocatedHandle(string handleType, u64 handle)
 extern void vkErrInvalidDescriptorArrayElement(u64 set, u32 binding, u32 arrayIndex)
+extern void vkErrImageLayout(VkImageLayout layout, VkImageLayout expectedLayout)
```
I'm fine with this for now.

AWoloszyn commented on gapis/api/vulkan/extensions/khr_swapchain.api:
```diff
@@ -266,7 +266,11 @@ cmd VkResult vkQueuePresentKHR(
+      if (Semaphores[wait_semaphores[i]].Signaled) {
+        Semaphores[wait_semaphores[i]].Signaled = false
+      } else {
+        // TODO: handle case when semaphore is not yet signaled
```
This is going to be a weird, (and quite rare) case. In fact since no application I have ever seen uses events, I would posit that we will never actually hit this edge-case.

However we definitely should still handle it.
Would you mind adding a comment here describing the case in some detail (alternatively, opening a github issue, and linking to it).

The only "legal" time this can happen is if we signal a semaphore after a VkCmdWaitForEvents, and then submit this before the associated VkCmdSetEvents/VkSetEvent is hit. If you can see any other legal cases, please also include them (and let me know)

bjoeris replied:
Added bug (https://github.com/google/gapid/issues/1860) and link in the TODO comment.

bjoeris commented on gapis/api/vulkan/api/queue.api:
```diff
@@ -109,6 +110,9 @@ cmd VkResult vkQueueSubmit(
       enterSubcontext()
       cb := CommandBuffers[command_buffers[j]]
+      if cb.Recording != COMPLETED {
+        vkErrorInvalidCommandBuffer(command_buffers[j])
```
Added vkErrorCommandBufferIncomplete for this error.

bjoeris commented:
A few of my changes had gotten lost in rebasing before initially submitting the pull request. In addition to addressing the comments, the latest version of this pull request made changes in the following places (compared to the version which was reviewed):
- coherent_memory.api: 
  - readmemoryInBufferBindings
  - writeMemoryInBufferBindings
  - readWriteMemoryInBoundDescriptorSets
- properties_features_requirements.api
  - vkGetBufferMemoryRequirements
- renderpass_framebuffer.api
  - dovkCmdBeginRenderPass
  - dovkCmdNextSubpass
  - loadImageAttachment

AWoloszyn commented on gapii/cc/vulkan_inlines.inc:
```diff
@@ -38,3 +38,11 @@ inline void VulkanSpy::vkErrExpectNVDedicatedlyAllocatedHandle(CallObserver*, st
 }
+
+inline void VulkanSpy::vkErrCommandBufferIncomplete(CallObserver*, VkCommandBuffer cmdbuf) {
+    GAPID_WARNING("Error: Executing command buffer %llu was not in the COMPLETED", cmdbuf)
```
not in the COMPLETED state"

bjoeris replied:
Fixed

AWoloszyn commented on gapis/api/vulkan/api/draw_commands.api:
```diff
@@ -94,7 +94,7 @@ class CmdBindBuffer {
 sub void dovkCmdBindVertexBuffers(ref!vkCmdBindVertexBuffersArgs bind) {
   for i , _ , v in bind.Buffers {
-    lastDrawInfo().BoundVertexBuffers[as!u32(i)] = BoundBuffer(
+    lastDrawInfo().BoundVertexBuffers[as!u32(i) + bind.FirstBinding] = BoundBuffer(
```
Good catch

AWoloszyn commented on gapis/api/vulkan/api/queue.api:
```diff
@@ -163,7 +170,11 @@ cmd VkResult vkQueueSubmit(
   execPendingCommands(queue)
-  fence
+  fence // 'fence' keyword, marking the point where observed memory writes become visible
+  if (fence != 0) { // 'fence' parameter, unrelated to 'fence' keyword above
```
Thanks for the comments

Qining commented:
Looks good to me

ben-clayton commented:
Will take a final pass look after lunch.

ben-clayton commented on gapis/api/vulkan/api/coherent_memory.api:
```diff
@@ -56,58 +56,170 @@ sub void readCoherentMemory(ref!DeviceMemoryObject memory, VkDeviceSize readOffs
-sub void readCoherentMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings) {
-  for _, _, v in bufferBindings {
+sub void readMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings, map!(u32, VkDeviceSize) bufferBindingOffsets) {
+  for k, _, v in bufferBindings {
```
Not sure if the logic is wrong here, but the triple is index, key, value. `k` is an odd variable choice for index, unless `k` always equals `i`.

Qining replied:
I think @ben-clayton caught a very important point! It looks like what we want here is `descriptor array index`, so we should use `for _, k, v in bufferBindings {`.

I think there may be similar spots around? And I think we should try it through all the samples and tests we have?

bjoeris replied:
The `k` name is due to an earlier version in which the loops in `readWriteMemoryInBoundDescriptorSets` (`i`), `readMemoryInDescriptorSet` (`j`), and `readMemoryInBufferBindings` (`k`) were nested range loops, similar to `dovkCmdBindDescriptorSets`.

I think you're right that using the map iteration index is wrong here. Since this uses Go maps under the hood, the iteration order is undefined, so I think the the index `k` is uncorrelated with the value `v`, correct?

I will switch these cases to `for i in (0 .. len(bufferBindings))` to match the pattern in `dovkCmdBindDescriptorSets` and emphasize that the keys are contiguous.

bjoeris replied:
I found what appear to be similar issues in `dovkCmdBindDescriptorSets`
```
  for i , _ , set in args.DescriptorSets {
    lastDrawInfo().DescriptorSets[args.FirstSet + as!u32(i)] = DescriptorSets[set]
  }
```
and `dovkCmdBindVertexBuffers`
```
  for i , _ , v in bind.Buffers ...
```
In each of these cases, it looks like the index `i` and the value, `set` or `v`, each range over the correct set, but in uncorrelated order, yes? If so I will include a fix of these together with the instances I introduced.

ben-clayton commented on gapis/api/vulkan/api/coherent_memory.api:
```diff
@@ -56,58 +56,170 @@ sub void readCoherentMemory(ref!DeviceMemoryObject memory, VkDeviceSize readOffs
+}
+
+sub void writeMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings, map!(u32, VkDeviceSize) bufferBindingOffsets) {
+  for k, _, v in bufferBindings {
```
ditto

ben-clayton commented on gapis/api/vulkan/api/renderpass_framebuffer.api:
```diff
@@ -210,39 +210,16 @@ vkCmdBeginRenderPassArgs {
   lastDrawInfo().RenderPass = RenderPasses[args.RenderPass]
   lastDrawInfo().InRenderPass = true
-  for _ , _ , v in lastDrawInfo().Framebuffer.ImageAttachments {
+  for _ , i , v in lastDrawInfo().Framebuffer.ImageAttachments {
```
humm...

bjoeris replied:
The keys in `ImageAttachments` are contiguous in the range `(0 .. let(ImageAttachments))` (see `vkCreateFramebuffer`). I think I will change these cases to use `for i in (0 .. len(imageAttachments))`, if there are no objections.

ben-clayton commented on gapis/api/vulkan/api/renderpass_framebuffer.api:
```diff
@@ -321,8 +302,8 @@ vkCmdEndRenderPassArgs {
 sub void dovkCmdEndRenderPass(ref!vkCmdEndRenderPassArgs unused) {
-  for _ , i , a in lastDrawInfo().RenderPass.AttachmentDescriptions {
-    lastDrawInfo().Framebuffer.ImageAttachments[i].Image.Info.Layout = a.finalLayout
+  for _ , i , _ in lastDrawInfo().RenderPass.AttachmentDescriptions {
```
again, `i` vs `k`

Qining commented on gapis/api/vulkan/api/coherent_memory.api:
```diff
@@ -56,58 +56,170 @@ sub void readCoherentMemory(ref!DeviceMemoryObject memory, VkDeviceSize readOffs
         VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC,
         VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC:
-          readCoherentMemoryInBufferBindings(v.BufferBinding)
+          readMemoryInBufferBindings(v.BufferBinding, bufferBindingOffsets[as!u32(j)])
```
Looks like we want `j` to be the binding number here, but actually `j` is index. Probably we should use `for _, j, v in descriptor_set.Bindings {`?

AWoloszyn commented:
I can't actually find this comment in the  MR any more, but I try
responding via email.

> The keys in ImageAttachments are contiguous in the range (0 ..
> let(ImageAttachments)) (see vkCreateFramebuffer). I think I will change
> these cases to use for i in (0 .. len(imageAttachments)), if there are no
> objections.
>
 We make this assumption in a lot of places. The api language has no vector
or similar, so we use this pattern in a number of places.


On Mon, May 7, 2018 at 11:48 AM Benson Joeris <notifications@github.com>
wrote:

> *@bjoeris* commented on this pull request.
> ------------------------------
>
> In gapis/api/vulkan/api/renderpass_framebuffer.api
> <https://github.com/google/gapid/pull/1845#discussion_r186464197>:
>
> >    lastDrawInfo().Framebuffer = Framebuffers[args.Framebuffer]
>    lastDrawInfo().LastSubpass = 0
>    lastDrawInfo().RenderPass = RenderPasses[args.RenderPass]
>    lastDrawInfo().InRenderPass = true
> -  for _ , _ , v in lastDrawInfo().Framebuffer.ImageAttachments {
> +  for _ , i , v in lastDrawInfo().Framebuffer.ImageAttachments {
>
> The keys in ImageAttachments are contiguous in the range (0 ..
> let(ImageAttachments)) (see vkCreateFramebuffer). I think I will change
> these cases to use for i in (0 .. len(imageAttachments)), if there are no
> objections.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/gapid/pull/1845#discussion_r186464197>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AIn9DJ9i4v6MGxuIBEI6pQQGBBGEptglks5twGyygaJpZM4TuMrM>
> .
>

bjoeris commented:
I have manually tested capture and replay in the UI with several of the SaschaWillems samples on Linux. I have not done any Android testing. The only issues I found were:
1. Mid-execution capture of shadowmappingcascade contain no commands: both the UI and `gapit commands` show that the capture file is empty. However, this behaviour is also present in master.
2. Mid-execution captures made with other gapid versions cannot be loaded by this version, because the state structure has changed.

Qining commented on gapis/api/vulkan/api/descriptor.api:
```diff
@@ -618,25 +618,41 @@ vkCmdBindDescriptorSetsArgs {
       // and see if it is a buffer.
-      for j in (0 .. len(set.Bindings)) {
-        binding := set.Bindings[as!u32(j)]
+      for j in (0 .. len(setObj.Bindings)) {
```
`setObj.Bindings` is a map of binding number to binding content, not a list of bindings. So for a descriptor set (actually defined in descriptor set layout), the valid bindings might be number 1, 2, 5, for example, then using the `j` here with list looping index: 0, 1, 2, is not correct.
